### PR TITLE
[Syntax] support nul character as garbage text trivia in libSyntax

### PR DESCRIPTION
**Pull request by omochi**
In previously, Lexer does not handle nul character in `lexTrivia` function.
So libSyntax lose information about nul character and can not achieve round trip translation.

With this PR, I update `lexTrivia` and libSyntax handles nul character as garbage text trivia.

I split my work into 2 commits.

In first commit, I introduce `Lexer::NulCharacterKind` enum to refactor Lexer code.
There are 3 types of meaning about nul character in Lexer.

- Embedded nul character
- CodeCompletion marker
- String buffer terminator

We need to handle correctly and carefully these cases when see nul character in Lexer.
To clear this task and prevent to forget about this, this enum is useful.

How do you think?

In second commit, I update `lexTrivia`.
This commit achieve main purpose of this PR.
And add test cases to check round trip of source file with nul, trivia construction, diagnostics about embedded nul character.

A new test case `Syntax/tokens_nul.swift` covers all consideration of `Syntax/lexer_invalid_nul.swift` which is added by me in previous PR https://github.com/apple/swift/pull/14954
So I removed this.

### CODE REVIEW

File: include/swift/Parse/Lexer.h [hunk 1]
```diff
@@ -461,6 +461,9 @@ class Lexer {
   };
 
 private:
+  /// nul character meaning kind
```
Comment (xwu) [inline include/swift/Parse/Lexer.h]: Drive-by nit: Please capitalize and punctuate comments.

**Comment by omochi** [comment]
@xwu Thanks, I updated.

File: lib/Parse/Lexer.cpp [hunk 2]
```diff
@@ -1857,6 +1869,16 @@ bool Lexer::tryLexConflictMarker(bool EatNewline) {
   return false;
 }
 
+Lexer::NulCharacterKind Lexer::getNulCharacterKind(const char *Ptr) const {
+  assert(Ptr != nullptr);
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: Could you add `&& *Ptr == 0`?

File: include/swift/Parse/Lexer.h [hunk 3]
```diff
@@ -461,7 +461,11 @@ class Lexer {
   };
 
 private:
-  /// nul character meaning kind
+  /// Nul character meaning kind.
+  /// There are 3 kinds of meaning about nul character in Lexer.
+  /// 1. BufferEnd is string buffer terminator.
+  /// 2. Embedded is embedded nul character.
+  /// 3. CodeCompletion is code completion marker.
   enum class NulCharacterKind { BufferEnd, Embedded, CodeCompletion };
```
Comment (rintaro) [inline include/swift/Parse/Lexer.h]: Please make this:
```cpp
/// Nul character meaning kind.
enum class NulCharacterKind {
  /// String buffer terminator.
  BufferEnd,
  /// Embedded is embedded nul character.
  Embedded,
...
```
numbered list (1,2,3) might be confusable with step-by-step procedure.

File: lib/Parse/Lexer.cpp (hunk 2 above)
Comment (omochi) [reply lib/Parse/Lexer.cpp]: Oh, good idea. I will fix it.

File: include/swift/Parse/Lexer.h (hunk 3 above)
Comment (omochi) [reply include/swift/Parse/Lexer.h]: Okey, I will fix it.

**Comment by omochi** [comment]
Ah, I mistook to apply changes to correct commit.
I will fix history again.

**Comment by omochi** [comment]
Ok, please review it.

File: lib/Parse/Lexer.cpp [hunk 4]
```diff
@@ -353,8 +354,13 @@ void Lexer::skipToEndOfLine(bool EatNewline) {
     case 0:
       // If this is a random nul character in the middle of a buffer, skip it as
       // whitespace.
-      if (CurPtr-1 != BufferEnd) {
+      switch (getNulCharacterKind(CurPtr - 1)) {
+      case NulCharacterKind::Embedded:
         diagnoseEmbeddedNul(Diags, CurPtr-1);
+        LLVM_FALLTHROUGH;
+      case NulCharacterKind::CodeCompletion:
+        goto LoopStart;
```
Comment (harlanhaskins) [inline lib/Parse/Lexer.cpp]: Can this be replaced with a `continue`?

File: lib/Parse/Lexer.cpp [hunk 5]
```diff
@@ -424,8 +431,13 @@ void Lexer::skipSlashStarComment() {
...
       // If this is a random nul character in the middle of a buffer, skip it as
       // whitespace.
-      if (CurPtr-1 != BufferEnd) {
-        diagnoseEmbeddedNul(Diags, CurPtr-1);
+      switch (getNulCharacterKind(CurPtr - 1)) {
+      case NulCharacterKind::Embedded:
+        diagnoseEmbeddedNul(Diags, CurPtr - 1);
+        LLVM_FALLTHROUGH;
+      case NulCharacterKind::CodeCompletion:
+        goto LoopStart;
```
Comment (harlanhaskins) [inline lib/Parse/Lexer.cpp]: Same as above

File: lib/Parse/Lexer.cpp (hunk 4 above)
Comment (omochi) [reply lib/Parse/Lexer.cpp]: Oh, thanks.
I overlooked `continue` in `switch` is bound to outer `while` not to `switch` what is different from `break`.
Interesting...
I will fix it so.

File: lib/Parse/Lexer.cpp (hunk 5 above)
Comment (omochi) [reply lib/Parse/Lexer.cpp]: OK

**Comment by omochi** [comment]
I updated in response to review.

File: lib/Parse/Lexer.cpp [hunk 6]
```diff
@@ -353,8 +353,13 @@ void Lexer::skipToEndOfLine(bool EatNewline) {
     case 0:
       // If this is a random nul character in the middle of a buffer, skip it as
       // whitespace.
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: This comment should be moved into `case NulCharacterKind::Embedded:` branch.

File: lib/Parse/Lexer.cpp [hunk 7]
```diff
@@ -353,8 +353,13 @@ void Lexer::skipToEndOfLine(bool EatNewline) {
...
-      if (CurPtr-1 != BufferEnd) {
+      switch (getNulCharacterKind(CurPtr - 1)) {
+      case NulCharacterKind::Embedded:
         diagnoseEmbeddedNul(Diags, CurPtr-1);
+        LLVM_FALLTHROUGH;
+      case NulCharacterKind::CodeCompletion:
+        continue;
+      case NulCharacterKind::BufferEnd:
         break;
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: Could you move the task in the case of `NulCharacterKind::BufferEnd` into this branch? i.e.
```cpp
case NulCharacterKind::BufferEnd:
  // The last line of the file does not have a newline.
  --CurPtr;
  return;
}
```

File: lib/Parse/Lexer.cpp [hunk 8]
```diff
@@ -424,8 +429,13 @@ void Lexer::skipSlashStarComment() {
...
-      if (CurPtr-1 != BufferEnd) {
-        diagnoseEmbeddedNul(Diags, CurPtr-1);
+      switch (getNulCharacterKind(CurPtr - 1)) {
+      case NulCharacterKind::Embedded:
+        diagnoseEmbeddedNul(Diags, CurPtr - 1);
+        LLVM_FALLTHROUGH;
+      case NulCharacterKind::CodeCompletion:
+        continue;
+      case NulCharacterKind::BufferEnd:
         break;
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: ditto

File: lib/Parse/Lexer.cpp [hunk 9]
```diff
@@ -2164,14 +2184,17 @@ void Lexer::lexImpl() {
...
-    // If this is a random nul character in the middle of a buffer, skip it as
-    // whitespace.
-    if (CurPtr-1 != BufferEnd) {
+    case NulCharacterKind::Embedded:
+      // If this is a random nul character in the middle of a buffer, skip it as
+      // whitespace.
       diagnoseEmbeddedNul(Diags, CurPtr-1);
       goto Restart;
+    case NulCharacterKind::BufferEnd:
+      break;
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: ditto

**Comment by omochi** [comment]
I updated.
@rintaro Thanks, I intended to reduce diff at first time. But I agree your review.

File: lib/Parse/Lexer.cpp [hunk 10]
```diff
@@ -422,26 +425,29 @@ void Lexer::skipSlashStarComment() {
...
-      --CurPtr;
+      switch (getNulCharacterKind(CurPtr - 1)) {
+      case NulCharacterKind::Embedded:
+        // If this is a random nul character in the middle of a buffer, skip it
+        // as whitespace.
+        diagnoseEmbeddedNul(Diags, CurPtr - 1);
+        LLVM_FALLTHROUGH;
+      case NulCharacterKind::CodeCompletion:
+        continue;
+      case NulCharacterKind::BufferEnd:
```
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: The last request from me.
Please enclose this branch with brace `case NulCharacterKind::BufferEnd: { ... }`. Lacking braces make us difficult to add another `case` (or re-order `case`s).

**Comment by omochi** [comment]
updated 🙂

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by rintaro — APPROVED** [review]
